README.md
# Circular Page View 

 A new Flutter package for Android, iOS & Web which helps in creating animated circular carousel with configurable settings.

### Preview

![](./assets/preview.gif)


### Usage

- Import the package
```
dependencies:
  circular_page_view: ^0.0.1
```

- Import in the file

```dart
import 'package:circular_page_view/circular_page_view.dart';
```

- Use of the Widget

```dart
CircularPageView(
  itemCount: 8,
  controller: _pageController,
  items: List.generate(8, (index) => _buildPage(index)),
  onPageChanged: (value) => setState(() => _currentPage = value),
  innerRadius: 0.8,
  offset: 150,
),
```



CHANGELOG.md
## [0.0.2] - 11th October 2020

* Minor Bug Fixes
## [0.0.1] - 10th October 2020

* A package to create animated circular carousel


test/circular_page_view_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:circular_page_view/circular_page_view.dart';

void main() {
}


lib/circular_page_view.dart
library circular_page_view;

import 'package:flutter/material.dart';

/// A PageView which creates Circular Scroll Effect.
class CircularPageView extends StatefulWidget {
  final PageController controller;
  final int itemCount;
  final List<Widget> items;

  ///Function to access the PageChanged status and returns currently Active Index
  final Function(int) onPageChanged;

  ///Value to control the amount of circular effect you want to create.
  ///It accepts value between 0.0 and 1.0
  final double innerRadius;

  ///Value to control the amount of offset you want the widgets to move down
  ///Best values range (100 - 300)
  final double offset;

  CircularPageView({
    @required this.controller,
    @required this.itemCount,
    @required this.items,
    this.onPageChanged,
    this.innerRadius = 0.5,
    this.offset = 150.0,
  });

  @override
  _CircularPageViewState createState() => _CircularPageViewState();
}

class _CircularPageViewState extends State<CircularPageView> {
  @override
  Widget build(BuildContext context) {
    return PageView.builder(
      controller: widget.controller,
      itemCount: widget.itemCount,
      itemBuilder: (context, index) => _buildPage(
        index,
        widget.items[index],
      ),
      onPageChanged: widget.onPageChanged,
    );
  }

  Widget _buildPage(int index, Widget mainChild) {
    return AnimatedBuilder(
      animation: widget.controller,
      builder: (BuildContext context, Widget child) {
        double value = 1;
        if (widget.controller.position.haveDimensions) {
          value = widget.controller.page - index;
          value = (1 - (value.abs() * widget.innerRadius)).clamp(0.0, 1.0);
          return Align(
            alignment: Alignment.center,
            child: Transform.translate(
              offset: Offset(
                0,
                -(Curves.ease.transform(value) * widget.offset),
              ),
              child: child,
            ),
          );
        } else {
          return Align(
            alignment: Alignment.center,
            child: Transform.translate(
              offset: Offset(
                0.0,
                -(Curves.ease.transform(index == 0 ? value : value * widget.innerRadius) * widget.offset),
              ),
              child: child,
            ),
          );
        }
      },
      child: mainChild,
    );
  }
}


example/lib/home_screen.dart
import 'package:circular_page_view/circular_page_view.dart';
import 'package:flutter/material.dart';

class HomeScreen extends StatefulWidget {
  @override
  _HomeScreenState createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  PageController _pageController;

  final Duration _animationDuration = Duration(milliseconds: 500);
  final Curve _animationCurve = Curves.easeInOut;

  @override
  void initState() {
    super.initState();

    _pageController = PageController(
      initialPage: 0,
      viewportFraction: 0.5,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Circular Page View'),
      ),
      body: Column(
        children: [
          SizedBox(
            height: MediaQuery.of(context).size.height * 0.7,
            child: CircularPageView(
              itemCount: 8,
              controller: _pageController,
              items: List.generate(8, (index) => _buildPage(index)),
              innerRadius: 0.7,
              offset: 180,
            ),
          ),
          SizedBox(
            width: MediaQuery.of(context).size.width * 0.5,
            height: MediaQuery.of(context).size.height * 0.1,
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                //previous button
                IconButton(
                  icon: Icon(
                    Icons.arrow_back_ios,
                    color: Theme.of(context).primaryColor,
                  ),
                  onPressed: () => _pageController.previousPage(duration: _animationDuration, curve: _animationCurve),
                ),
                //next button
                IconButton(
                  icon: Icon(
                    Icons.arrow_forward_ios,
                    color: Theme.of(context).primaryColor,
                  ),
                  onPressed: () => _pageController.nextPage(duration: _animationDuration, curve: _animationCurve),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildPage(int index) {
    return Container(
      height: 150,
      width: 150,
      margin: const EdgeInsets.all(22.0),
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(100.0),
        color: Colors.red,
      ),
      alignment: Alignment.center,
      child: Text(
        '$index',
        style: TextStyle(color: Colors.white, fontSize: 28.0, fontWeight: FontWeight.bold),
      ),
    );
  }
}


example/circular_page_example.dart
import 'package:flutter/material.dart';
import 'package:circular_page_view/circular_page_view.dart';

class HomeScreen extends StatefulWidget {
  @override
  _HomeScreenState createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  PageController _pageController;

  int _currentPage = 0;

  final Duration _animationDuration = Duration(milliseconds: 500);
  final Curve _animationCurve = Curves.easeInOut;

  @override
  void initState() {
    super.initState();

    _pageController = PageController(
      initialPage: 0,
      viewportFraction: 0.5,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Circular Page View'),
      ),
      body: Column(
        children: [
          SizedBox(
            height: MediaQuery.of(context).size.height * 0.8,
            child: CircularPageView(
              itemCount: 8,
              controller: _pageController,
              items: List.generate(8, (index) => _buildPage(index)),
              onPageChanged: (value) => setState(() => _currentPage = value),
              innerRadius: 0.8,
              offset: 150,
            ),
          ),
          SizedBox(
            width: MediaQuery.of(context).size.width * 0.5,
            height: MediaQuery.of(context).size.height * 0.1,
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                //previous button
                IconButton(
                  icon: Icon(
                    Icons.arrow_back_ios,
                    color: Theme.of(context).primaryColor,
                  ),
                  onPressed: () => _pageController.previousPage(duration: _animationDuration, curve: _animationCurve),
                ),
                //next button
                IconButton(
                  icon: Icon(
                    Icons.arrow_forward_ios,
                    color: Theme.of(context).primaryColor,
                  ),
                  onPressed: () => _pageController.nextPage(duration: _animationDuration, curve: _animationCurve),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildPage(int index) {
    return Container(
      height: 150,
      width: 150,
      margin: const EdgeInsets.all(22.0),
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(100.0),
        color: Colors.red,
      ),
      alignment: Alignment.center,
      child: Text(
        '$index',
        style: TextStyle(color: Colors.white, fontSize: 28.0, fontWeight: FontWeight.bold),
      ),
    );
  }
}
